Feat/assets masks

wildintellect opened the pull request:
Before you submit a pull request, please fill in the following:

**Related Issue(s):**
#6 

**Description:**
Adding the additional assets that are available. See #6 for more details.

**PR checklist:**

- [x] Code is formatted (run `scripts/format`).
- [x] Code lints properly (run `scripts/lint`).
- [x] Tests pass (run `scripts/test`).
- [ ] Documentation has been updated to reflect changes, if applicable.
- [x] Changes are added to the [CHANGELOG](../CHANGELOG.md).

wildintellect committed:
feat: Add additional assets

TODO: fix paths to assets, put correct media type for kml

wildintellect committed:
Merge branch 'main' into feat/assets_masks

wildintellect committed:
fix: Asset paths based on initial asset href

wildintellect committed:
feat: Correct href to assets in AUX and Previews

TODO: inclusion of additional assets should probably be an optional arg

wildintellect committed:
chore: cleanup of prior method

wildintellect committed:
chore: forgot to add asset_utils, PREVIEW not PREVIEWS

wildintellect committed:
chore: Typing typo

wildintellect committed:
chore: KML is more specific than XML

philvarner commented on examples/cop-dem-glo-30/Copernicus_DSM_COG_10_N53_00_W115_00_DEM/Copernicus_DSM_COG_10_N53_00_W115_00_DEM.json:
```diff
@@ -113,6 +124,129 @@
         "data"
       ]
+    },
+    "SRC": {
```
I can't recall if this is in the STAC spec or not, but I think it's more conventional to use lowercase asset names here.

wildintellect replied:
Seems to be convention, not part of the spec, particularly inherited from json. Was a little funky but found a solution.

wildintellect committed:
fix: asset names as lowercase by convention

wildintellect committed:
chore: Update changelog with new Asset info.

wildintellect commented:
@philvarner do you think we need the Bands info on all the new assets (masks and previews)?

chuckwondo commented on examples/cop-dem-glo-30/Copernicus_DSM_COG_10_N53_00_W115_00_DEM/Copernicus_DSM_COG_10_N53_00_W115_00_DEM.json:
```diff
@@ -113,6 +124,129 @@
+      "href": "../../tests/data-files/external/PREVIEW/Copernicus_DSM_COG_10_N53_00_W115_00_DEM_QL.tif",
+      "type": "image/tiff",
+      "title": "DEM QuickLook Relative",
+      "description": "Quicklook of Digital Elevation Model wiuth shaded relief representation (relative)",
```
```suggestion
      "description": "QuickLook of Digital Elevation Model with shaded relief representation (relative)",
```

chuckwondo commented on examples/cop-dem-glo-30/Copernicus_DSM_COG_10_N53_00_W115_00_DEM/Copernicus_DSM_COG_10_N53_00_W115_00_DEM.json:
```diff
@@ -113,6 +124,129 @@
+      "href": "../../tests/data-files/external/PREVIEW/Copernicus_DSM_COG_10_N53_00_W115_00_QL.kml",
+      "type": "application/vnd.google-earth.kml+xml",
+      "title": "QuickLook kml",
+      "description": "kml fo QuickLook data visualization",
```
```suggestion
      "description": "kml of QuickLook data visualization",
```

chuckwondo commented on examples/cop-dem-glo-30/Copernicus_DSM_COG_10_N53_00_W115_00_DEM/Copernicus_DSM_COG_10_N53_00_W115_00_DEM.json:
```diff
@@ -113,6 +124,129 @@
+      "href": "../../tests/data-files/external/PREVIEW/Copernicus_DSM_COG_10_N53_00_W115_00_DEM_ABS_QL.tif",
+      "type": "image/tiff",
+      "title": "DEM QuickLook Absolute",
+      "description": "Quicklook of Digital Elevation Model wiuth shaded relief representation (absolute)",
```
```suggestion
      "description": "QuickLook of Digital Elevation Model with shaded relief representation (absolute)",
```

chuckwondo commented on examples/cop-dem-glo-30/collection.json:
```diff
@@ -39,6 +45,116 @@
+    "dem_ql": {
+      "title": "DEM QuickLook Relative",
+      "type": "image/tiff",
+      "description": "Quicklook of Digital Elevation Model wiuth shaded relief representation (relative)",
```
```suggestion
      "description": "Quicklook of Digital Elevation Model with shaded relief representation (relative)",
```

chuckwondo commented on examples/cop-dem-glo-30/collection.json:
```diff
@@ -39,6 +45,116 @@
+    "ql": {
+      "title": "QuickLook kml",
+      "type": "application/vnd.google-earth.kml+xml",
+      "description": "kml fo QuickLook data visualization",
```
```suggestion
      "description": "kml of QuickLook data visualization",
```

chuckwondo commented on examples/cop-dem-glo-30/collection.json:
```diff
@@ -39,6 +45,116 @@
+    "dem_abs_ql": {
+      "title": "DEM QuickLook Absolute",
+      "type": "image/tiff",
+      "description": "Quicklook of Digital Elevation Model wiuth shaded relief representation (absolute)",
```
```suggestion
      "description": "QuickLook of Digital Elevation Model with shaded relief representation (absolute)",
```

chuckwondo commented on examples/cop-dem-glo-90/Copernicus_DSM_COG_30_N53_00_W115_00_DEM/Copernicus_DSM_COG_30_N53_00_W115_00_DEM.json:
```diff
@@ -113,6 +124,129 @@
+      "href": "../../tests/data-files/external/PREVIEW/Copernicus_DSM_COG_30_N53_00_W115_00_DEM_QL.tif",
+      "type": "image/tiff",
+      "title": "DEM QuickLook Relative",
+      "description": "Quicklook of Digital Elevation Model wiuth shaded relief representation (relative)",
```
```suggestion
      "description": "QuickLook of Digital Elevation Model with shaded relief representation (relative)",
```

chuckwondo commented on examples/cop-dem-glo-90/Copernicus_DSM_COG_30_N53_00_W115_00_DEM/Copernicus_DSM_COG_30_N53_00_W115_00_DEM.json:
```diff
@@ -113,6 +124,129 @@
+      "href": "../../tests/data-files/external/PREVIEW/Copernicus_DSM_COG_30_N53_00_W115_00_QL.kml",
+      "type": "application/vnd.google-earth.kml+xml",
+      "title": "QuickLook kml",
+      "description": "kml fo QuickLook data visualization",
```
```suggestion
      "description": "kml of QuickLook data visualization",
```

chuckwondo commented on examples/cop-dem-glo-90/Copernicus_DSM_COG_30_N53_00_W115_00_DEM/Copernicus_DSM_COG_30_N53_00_W115_00_DEM.json:
```diff
@@ -113,6 +124,129 @@
+      "href": "../../tests/data-files/external/PREVIEW/Copernicus_DSM_COG_30_N53_00_W115_00_DEM_ABS_QL.tif",
+      "type": "image/tiff",
+      "title": "DEM QuickLook Absolute",
+      "description": "Quicklook of Digital Elevation Model wiuth shaded relief representation (absolute)",
```
```suggestion
      "description": "QuickLook of Digital Elevation Model with shaded relief representation (absolute)",
```

chuckwondo commented on examples/cop-dem-glo-90/collection.json:
```diff
@@ -39,6 +45,116 @@
+    "dem_ql": {
+      "title": "DEM QuickLook Relative",
+      "type": "image/tiff",
+      "description": "Quicklook of Digital Elevation Model wiuth shaded relief representation (relative)",
```
```suggestion
      "description": "QuickLook of Digital Elevation Model with shaded relief representation (relative)",
```

chuckwondo commented on examples/cop-dem-glo-90/collection.json:
```diff
@@ -39,6 +45,116 @@
+    "ql": {
+      "title": "QuickLook kml",
+      "type": "application/vnd.google-earth.kml+xml",
+      "description": "kml fo QuickLook data visualization",
```
```suggestion
      "description": "kml of QuickLook data visualization",
```

chuckwondo commented on examples/cop-dem-glo-90/collection.json:
```diff
@@ -39,6 +45,116 @@
+    "dem_abs_ql": {
+      "title": "DEM QuickLook Absolute",
+      "type": "image/tiff",
+      "description": "Quicklook of Digital Elevation Model wiuth shaded relief representation (absolute)",
```
```suggestion
      "description": "QuickLook of Digital Elevation Model with shaded relief representation (absolute)",
```

chuckwondo commented on src/stactools/cop_dem/assets_utils.py:
```diff
@@ -0,0 +1,18 @@
+    """
+    Apply a subdirectory prefix, and ensure the correct capitalization of files.
+    """
+    asset_href = os.path.join(os.path.dirname(href), subdirectory,
+                              os.path.basename(href))
```
I suggest avoiding using `os.path` functions when dealing with URLs, as these have OS-dependent behavior. In particular, Windows uses backslashes (`\`) in paths.

Since I'm not particularly familiar with what you're trying to achieve here, I might be off, but in the STAC spec, the `href` of an asset is a URI, and URIs are OS-independent, so you should not be using functions that have OS-dependent behavior to deal with them.

Further, not all URIs are URLs, so a URI might not even contain slashes (`/`). However, if you know here that you'll be dealing only with URIs that are URLs, then you may be fine here.

If you know you'll only deal with URLs, then I would at least recommend using functions from `urllib.parse` in order to avoid OS-dependent behavior that won't work on all OSs:

```suggestion
    from urllib.parse import urljoin
    _, basename = url.rsplit('/', 1)
    asset_href = urljoin(href, f"{subdirectory}/{basename}")
```

To show that this works:

```
>>> href = "http://host/path/to/asset.ext"
>>> _, basename = href.rsplit("/", 1)
>>> subdirectory = "subdir"
>>> urljoin(href, f"{subdirectory}/{basename}")
'http://host/path/to/subdir/asset.ext'
>>> # Trailing slash is automatically removed
>>> subdirectory = "subdir/"
>>> urljoin(href, f"{subdirectory}/{basename}")
'http://host/path/to/subdir/asset.ext'
>>> # Leading slash allows full path replacement
>>> subdirectory = "/subdir"
>>> urljoin(href, f"{subdirectory}/{basename}")
'http://host/subdir/asset.ext'
```

philvarner replied:
The spec requires it to be a URL, and it's allowed to be a local filepath.

chuckwondo replied:
@philvarner, this is the spec I'm reading: https://github.com/radiantearth/stac-spec/blob/master/item-spec/item-spec.md#asset-object

Is that not the right spec?

philvarner replied:
I just read the same thing, I have have no idea why I read that as URL. You're right.

chuckwondo replied:
Regardless, a [file URI](https://en.wikipedia.org/wiki/File_URI_scheme) (URL) still uses forward slashes.

chuckwondo commented on src/stactools/cop_dem/assets_utils.py:
```diff
@@ -0,0 +1,18 @@
+from typing import Tuple
+
+
+def change_asset_directory(href: str, asset_name: str,
+                           subdirectory: str) -> Tuple[str, str]:
```
Since all 3 parameters are strings, it can be easy to specify arguments in the wrong order without being alerted to the problem. To help reduce the chances of this, you might want to consider kwargs-only parameters:
```suggestion
def change_asset_directory(*, href: str, asset_name: str,
                           subdirectory: str) -> Tuple[str, str]:
```

Then, when calling the function, the caller is forced to specify named arguments. While it would still be possible to mis-specify arguments, you at least stand a better chance of not doing so because you'll have an immediate visual clue that something is wrong).

```
# Not obvious that args 2 and 3 are swapped, and might
# even seem like a more logical argument order
change_asset_directory(href, subdir, asset)

# Order doesn't matter, and names make it less likely to
# supply the wrong values with the names
change_asset_directory(href=href, subdir=subdir, asset=asset)
```

chuckwondo reviewed (changes requested):
Primarily spelling errors, but there are a couple of code suggestions to consider.

philvarner commented:
> @philvarner do you think we need the Bands info on all the new assets (masks and previews)?

I think it might be useful on the masks. Previews generally don't have them.

philvarner commented on examples/cop-dem-glo-30/Copernicus_DSM_COG_10_N53_00_W115_00_DEM/Copernicus_DSM_COG_10_N53_00_W115_00_DEM.json:
```diff
@@ -113,6 +124,129 @@
+      "title": "Source mask",
+      "description": "Source Mask",
+      "role": [
+        "data",
```
I feel like this shouldn't have both data and data-mask, like it should be one or the other? maybe `mask` is a sufficient name for this  too?

wildintellect replied:
I'm finding other examples around that say "data-mask" but agree that "mask" should be sufficient.

philvarner commented on examples/cop-dem-glo-30/Copernicus_DSM_COG_10_N53_00_W115_00_DEM/Copernicus_DSM_COG_10_N53_00_W115_00_DEM.json:
```diff
@@ -113,6 +124,129 @@
+    },
+    "dem_ql": {
+      "href": "../../tests/data-files/external/PREVIEW/Copernicus_DSM_COG_10_N53_00_W115_00_DEM_QL.tif",
+      "type": "image/tiff",
```
are these any more specific than just tifs (geotiff or cog)?

wildintellect replied:
Looks like they are GeoTiff but not COG, I'll update.

philvarner commented on examples/cop-dem-glo-30/Copernicus_DSM_COG_10_N53_00_W115_00_DEM/Copernicus_DSM_COG_10_N53_00_W115_00_DEM.json:
```diff
@@ -113,6 +124,129 @@
+    "ql": {
+      "href": "../../tests/data-files/external/PREVIEW/Copernicus_DSM_COG_10_N53_00_W115_00_QL.kml",
+      "type": "application/vnd.google-earth.kml+xml",
+      "title": "QuickLook kml",
```
```suggestion
      "title": "QuickLook KML",
```

philvarner commented on examples/cop-dem-glo-30/Copernicus_DSM_COG_10_N53_00_W115_00_DEM/Copernicus_DSM_COG_10_N53_00_W115_00_DEM.json:
```diff
@@ -113,6 +124,129 @@
+    "acm": {
+      "href": "../../tests/data-files/external/AUXFILES/Copernicus_DSM_COG_10_N53_00_W115_00_ACM.kml",
+      "type": "application/vnd.google-earth.kml+xml",
+      "title": "Height Error mask",
```
I feel like the title should be all capitalized words, and the description should only have the first one capitalized (and maybe a bit more text after it).

wildintellect committed:
Update examples/cop-dem-glo-30/Copernicus_DSM_COG_10_N53_00_W115_00_DEM/Copernicus_DSM_COG_10_N53_00_W115_00_DEM.json

Co-authored-by: Chuck Daniels <cjdaniels4@gmail.com>

wildintellect committed:
Update examples/cop-dem-glo-30/Copernicus_DSM_COG_10_N53_00_W115_00_DEM/Copernicus_DSM_COG_10_N53_00_W115_00_DEM.json

Co-authored-by: Chuck Daniels <cjdaniels4@gmail.com>

wildintellect committed:
Update examples/cop-dem-glo-30/Copernicus_DSM_COG_10_N53_00_W115_00_DEM/Copernicus_DSM_COG_10_N53_00_W115_00_DEM.json

Co-authored-by: Chuck Daniels <cjdaniels4@gmail.com>

wildintellect committed:
Update examples/cop-dem-glo-30/collection.json

Co-authored-by: Chuck Daniels <cjdaniels4@gmail.com>

wildintellect committed:
Update examples/cop-dem-glo-30/collection.json

Co-authored-by: Chuck Daniels <cjdaniels4@gmail.com>

wildintellect committed:
Update examples/cop-dem-glo-90/collection.json

Co-authored-by: Chuck Daniels <cjdaniels4@gmail.com>

wildintellect committed:
Update examples/cop-dem-glo-90/collection.json

Co-authored-by: Chuck Daniels <cjdaniels4@gmail.com>

wildintellect committed:
chore: Typos

wildintellect committed:
fix: role 'mask' not data

jjfrench commented on src/stactools/cop_dem/constants.py:
```diff
@@ -55,17 +55,117 @@
+        "title": "DEM QuickLook Relative",
+        "type": MediaType.TIFF,
+        "description":
+        "Quicklook of Digital Elevation Model wiuth shaded relief representation (relative)",
```
```suggestion
        "QuickLook of Digital Elevation Model with shaded relief representation (relative)",
```
Based off Chuck's comments - should make the change here and rerun examples

jjfrench commented on src/stactools/cop_dem/constants.py:
```diff
@@ -55,17 +55,117 @@
+        "title": "DEM QuickLook Absolute",
+        "type": MediaType.TIFF,
+        "description":
+        "Quicklook of Digital Elevation Model wiuth shaded relief representation (absolute)",
```
```suggestion
        "QuickLook of Digital Elevation Model with shaded relief representation (absolute)",
```

wildintellect committed:
fix: resolve PR review comments #31, todo rasterbands

wildintellect committed:
chore: WIP classification for masks

wildintellect committed:
chore: classification for masks cleaned

wildintellect committed:
chore: classification for masks formatted

wildintellect committed:
feat: classes for mask assets

wildintellect committed:
feat: Hand coded raster:bands for masks

wildintellect commented:
@philvarner I added the data_type for the bands in the masks, nothing else seemed highly useful. Also realized the classes for those masks could be very useful so also added them but realized the [Classification Extension](https://github.com/stac-extensions/classification) isn't in pystac. Followed what [Landsat](https://github.com/stactools-packages/landsat/blob/main/src/stactools/landsat/stac.py) did in this regard.

chuckwondo commented on examples/cop-dem-glo-30/Copernicus_DSM_COG_10_N53_00_W115_00_DEM/Copernicus_DSM_COG_10_N53_00_W115_00_DEM.json:
```diff
@@ -73,6 +78,12 @@
+    {
+      "rel": "product handbook",
+      "href": "https://spacedata.copernicus.eu/documents/20123/122407/GEO1988-CopernicusDEM-SPE-002_ProductHandbook_I5.0+%281%29.pdf/706ee17d-2cce-f1fa-a73e-1686d28f09dd?t=1679657087883application/pdf",
+      "type": "Compernicus DEM Product Handbook (Nov 2022)",
```
```suggestion
      "type": "Copernicus DEM Product Handbook (Nov 2022)",
```

Perhaps also change "handbook" to "Handbook" on line 78 (a few lines above, for "title")

chuckwondo commented on examples/cop-dem-glo-30/Copernicus_DSM_COG_10_N53_00_W115_00_DEM/Copernicus_DSM_COG_10_N53_00_W115_00_DEM.json:
```diff
@@ -113,6 +124,304 @@
+    "ql": {
+      "href": "../../tests/data-files/external/PREVIEW/Copernicus_DSM_COG_10_N53_00_W115_00_QL.kml",
+      "type": "application/vnd.google-earth.kml+xml",
+      "title": "QuickLookKML",
```
```suggestion
      "title": "QuickLook KML",
```

chuckwondo commented on examples/cop-dem-glo-30/collection.json:
```diff
@@ -17,6 +17,12 @@
+    {
+      "rel": "product handbook",
+      "href": "https://spacedata.copernicus.eu/documents/20123/122407/GEO1988-CopernicusDEM-SPE-002_ProductHandbook_I5.0+%281%29.pdf/706ee17d-2cce-f1fa-a73e-1686d28f09dd?t=1679657087883application/pdf",
+      "type": "Compernicus DEM Product Handbook (Nov 2022)",
```
```suggestion
      "type": "Copernicus DEM Product Handbook (Nov 2022)",
```

chuckwondo commented on examples/cop-dem-glo-90/Copernicus_DSM_COG_30_N53_00_W115_00_DEM/Copernicus_DSM_COG_30_N53_00_W115_00_DEM.json:
```diff
@@ -73,6 +78,12 @@
+    {
+      "rel": "product handbook",
+      "href": "https://spacedata.copernicus.eu/documents/20123/122407/GEO1988-CopernicusDEM-SPE-002_ProductHandbook_I5.0+%281%29.pdf/706ee17d-2cce-f1fa-a73e-1686d28f09dd?t=1679657087883application/pdf",
+      "type": "Compernicus DEM Product Handbook (Nov 2022)",
```
```suggestion
      "type": "Copernicus DEM Product Handbook (Nov 2022)",
```

chuckwondo commented on examples/cop-dem-glo-90/collection.json:
```diff
@@ -17,6 +17,12 @@
+    {
+      "rel": "product handbook",
+      "href": "https://spacedata.copernicus.eu/documents/20123/122407/GEO1988-CopernicusDEM-SPE-002_ProductHandbook_I5.0+%281%29.pdf/706ee17d-2cce-f1fa-a73e-1686d28f09dd?t=1679657087883application/pdf",
+      "type": "Compernicus DEM Product Handbook (Nov 2022)",
```
```suggestion
      "type": "Copernicus DEM Product Handbook (Nov 2022)",
```

chuckwondo commented on src/stactools/cop_dem/constants.py:
```diff
@@ -55,17 +61,124 @@
+        "product handbook",
+        "https://spacedata.copernicus.eu/documents/20123/122407/GEO1988-CopernicusDEM-SPE-002_ProductHandbook_I5.0+%281%29.pdf/706ee17d-2cce-f1fa-a73e-1686d28f09dd?t=1679657087883"  # noqa: E501
+        "application/pdf",
+        "Compernicus DEM Product Handbook (Nov 2022)",
```
```suggestion
        "Copernicus DEM Product Handbook (Nov 2022)",
```

chuckwondo commented on examples/cop-dem-glo-30/collection.json:
```diff
@@ -39,6 +45,291 @@
+      ]
+    },
+    "ql": {
+      "title": "QuickLookKML",
```
```suggestion
      "title": "QuickLook KML",
```

chuckwondo commented on examples/cop-dem-glo-90/Copernicus_DSM_COG_30_N53_00_W115_00_DEM/Copernicus_DSM_COG_30_N53_00_W115_00_DEM.json:
```diff
@@ -113,6 +124,304 @@
+    "ql": {
+      "href": "../../tests/data-files/external/PREVIEW/Copernicus_DSM_COG_30_N53_00_W115_00_QL.kml",
+      "type": "application/vnd.google-earth.kml+xml",
+      "title": "QuickLookKML",
```
```suggestion
      "title": "QuickLook KML",
```

chuckwondo commented on examples/cop-dem-glo-90/collection.json:
```diff
@@ -39,6 +45,291 @@
+      ]
+    },
+    "ql": {
+      "title": "QuickLookKML",
```
```suggestion
      "title": "QuickLook KML",
```

chuckwondo commented on src/stactools/cop_dem/constants.py:
```diff
@@ -55,17 +61,124 @@
+    }),
+    "ql":
+    AssetDefinition({
+        "title": "QuickLookKML",
```
```suggestion
        "title": "QuickLook KML",
```

chuckwondo commented on src/stactools/cop_dem/stac.py:
```diff
@@ -108,6 +110,36 @@ def create_item(
+    # Add the AUX and Preview Assets together
+    assets_dict.update(assets_auxfiles_dict)
+
+    for key, value in assets_dict.items():
+        if (asset_def := co.COP_DEM_ASSETS.get(key)) is not None:
+            asset = asset_def.create_asset((value))
+
+        item.add_asset(key, asset)
```
```suggestion
    for key, value in assets_dict.items():
        if (asset_def := co.COP_DEM_ASSETS.get(key)) is not None:
            item.add_asset(key, asset_def.create_asset(value))
```

chuckwondo reviewed (changes requested):
A few more typos and a code fix.

philvarner commented on examples/cop-dem-glo-30/collection.json:
```diff
@@ -39,6 +45,291 @@
+          "data_type": "uint8"
+        }
+      ],
+      "classes": [
```
I think this should be `classification:classes`

wildintellect committed:
chore: typos fixes in #31

wildintellect commented:
- [x] @philvarner ah I see where I went wrong on the classification, with bitfields there's a nested "classes"
- [x] @chuckwondo fixed. FYI the typos all come from the constants.py which propagates to the json files when the examples are built.

chuckwondo reviewed (approved):
LGTM!

philvarner commented on examples/cop-dem-glo-30/Copernicus_DSM_COG_10_N53_00_W115_00_DEM/Copernicus_DSM_COG_10_N53_00_W115_00_DEM.json:
```diff
@@ -210,7 +210,7 @@
         }
       ],
-      "classes": [
+      "classification:classes": [
```
stac_extensions is missing Classification URI

philvarner commented on examples/cop-dem-glo-30/Copernicus_DSM_COG_10_N53_00_W115_00_DEM/Copernicus_DSM_COG_10_N53_00_W115_00_DEM.json:
```diff
@@ -122,7 +431,9 @@
   ],
   "stac_extensions": [
-    "https://stac-extensions.github.io/projection/v1.0.0/schema.json"
+    "https://stac-extensions.github.io/raster/v1.1.0/schema.json",
```
missing Classification URI

philvarner commented on src/stactools/cop_dem/stac.py:
```diff
@@ -108,6 +110,34 @@ def create_item(
     item.add_asset("data", data_asset)
     RasterExtension.ext(data_asset, add_if_missing=True).bands = [data_bands]
 
+    # Make a list of /AUXFILES assets
```
Classification URI should be added to stac_extensions here

wildintellect replied:
That's what I was asking about earlier when I said Pystac doesn't have the Classification ext, and I couldn't find any examples https://github.com/stactools-packages/landsat/blob/9f595a9d5ed6b62a2e96338e79f5bb502a7d90d0/examples/landsat-c2-l2/collection.json#L59-L64 uses it but doesn't list it anywhere. Is the only way to implement by adding it to Pystac as in this [tutorial](https://pystac.readthedocs.io/en/stable/tutorials/adding-new-and-custom-extensions.html)?

wildintellect replied:
I manually added the URI, does that work?

philvarner replied:
You do have to manually add the URI, but also to the Item (the code you added was only added to the Collection). 

I think this will work around line 154:

```
  item.stac_extensions.append("https://stac-extensions.github.io/classification/v1.1.0/schema.json")
```

Looks like this is what landsat is doing here     item.stac_extensions.append(CLASSIFICATION_EXTENSION_SCHEMA)

philvarner replied:
Also, actually implementing the ClassificationExtension would be a nice thing to have, and it doesn't have to be implemented in the pystac package. When I did the GridExtension, I just put it in stactools-sentinel2 and then ported it to stactools, and then when a newer version of stactools was released, replaced it it from that package with the pystac one.

wildintellect replied:
I see how Landsat did it now (though they missed listing it on the collection). Done.

https://github.com/stac-utils/pystac/issues/788 is already open for adding the extension.

philvarner replied:
Sounds good. I mostly wanted to make it clear that the extension doesn't **have** to be part of pystac.

wildintellect committed:
feat: add Classification URI manually, not in pystac

wildintellect committed:
fix: Add classification schema to items also

wildintellect commented:
@philvarner @gadomski @matthewhanson I don't have write access, so either need permissions or someone else has to push the merge button, thanks.
